modeling/algorithm/local-weighted-slice-algorithm.go

```diff
@@ -0,0 +1,211 @@
+		expectedEndpoints := zone.NodesRatio * float64(region.TotalEndpoints)
+		// deviation: a negative value means this zone needs more endpoints from
+		// other zones, a positive value means this zone needs to give out
+		// endpoints to other zones
+		deviation := float64(zone.Endpoints) - expectedEndpoints
+		// intDeviation is dealt same way as original local-slice algorithm that
+		// directly 'borrows' or 'sends' endpoints among zones
+		intDeviation := int(deviation)
+		if intDeviation == 0 {
+			localGroup.Composition[zoneName] = types.WeightedEndpoints{Number: int(expectedEndpoints), Weight: 1}
+		} else if intDeviation > 0 {
+			endpointsAvailable.push(endpointDeviation{name: zoneName, deviation: intDeviation})
+			localGroup.Composition[zoneName] = types.WeightedEndpoints{Number: int(expectedEndpoints), Weight: 1}
+		} else {
+			endpointsNeeded.push(endpointDeviation{name: zoneName, deviation: -intDeviation})
+			localGroup.Composition[zoneName] = types.WeightedEndpoints{Number: zone.Endpoints, Weight: 1}
+		}
+		sliceGroups[zoneName] = localGroup
+
+		// push 0.xx deviation into corresponding lists
+		// 2.3 expected endpoints, 3 ownd endpoints, 0.7 decimal deviation, 0.3
+		// endpoints and 2 endpoints needed
+		// 2.4 expected endpoints, 1 owned endpoint, -0.4 decimal deviation, 0.4
+		// endpoints and 2 endpoints needed
+		decimalDeviation := deviation - float64(intDeviation)
+		if decimalDeviation > 0 {
+			// One endpoint from this zone will contribute 1-decimalDeviation to
+			// the local zone, and contribute remaining to other zones. This
+			// will be implemented through routing weights of EndpointSliceGroup
+			weightedEndpointsAvailable.push(endpointDeviation{name: zoneName, deviation: 1, weight: 1 - decimalDeviation, consumeByLocal: true})
+		} else if decimalDeviation < 0 {
+			weightedEndpointsNeeded.push(endpointDeviation{name: zoneName, deviation: 1, weight: -decimalDeviation})
+		}
+	}
+
+	err := alg.balanceSliceGroups(&endpointsAvailable, &endpointsNeeded, &weightedEndpointsAvailable, &weightedEndpointsNeeded, sliceGroups)
+	return sliceGroups, err
+}
+
+// balanceSliceGroups distributes endpoints from zones with extra endpoints to
+// EndpointSliceGroups for zones with insufficient endpoints.
+func (alg LocalWeightedSliceAlgorithm) balanceSliceGroups(endpointsAvailable *endpointsList, endpointsNeeded *endpointsList, weightedEndpointsAvailable *endpointsList, weightedEndpointsNeeded *endpointsList, sliceGroups map[string]types.EndpointSliceGroup) error {
+
+	for _, receiveZone := range endpointsNeeded.byZone {
+		// There are no more full endpoints available, but this zone still needs
+		// endpoints. Push the needed endpoints into a weighted list and deal
+		// with them as partial endpoints.
+		if len(endpointsAvailable.byZone) == 0 {
+			receiveZone.weight = 1
+			weightedEndpointsNeeded.push(receiveZone)
+			endpointsNeeded.pop()
+			continue
+		}
+		// same logic as original local-slice algorithm
```

Comment:
If this is shared logic maybe we can extract it to a helper func?

Reply:
wrapped the logic to a helper function, same as the previous comment